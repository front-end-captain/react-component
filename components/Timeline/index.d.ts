import * as React from "react";

import TimelineItem from "./timelineItem";

interface TimelineProps {
  prefixCls?: string;
  className?: string;

  // 指定最后一个幽灵节点的内容或者是否存在
  pending?: React.ReactNode | boolean | string;

  // 当最后一个幽灵节点存在時，指定其时间图点
  pendingDot?: React.ReactNode;
  style?: React.CSSProperties;
  // reverse?: boolean;

  // 时间轴和内容的相对位置
  mode?: "left" | "alternate" | "right";
}

export default class Timeline extends React.Component<TimelineProps, any> {
  static Item: typeof TimelineItem;
}
