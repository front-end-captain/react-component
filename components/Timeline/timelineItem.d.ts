import * as React from "react";

interface TimelineItemProps {
  prefixCls?: string;
  className?: string;
  color?: string;
  dot?: React.ReactNode;
  pending?: boolean;
  style?: React.CSSProperties;
}

declare const TimelineItem: (props: TimelineItemProps) => React.ReactNode;

export default TimelineItem;
