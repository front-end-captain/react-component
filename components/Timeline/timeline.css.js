import styled from "styled-components";

import { skySpinRotate } from "./../_style/animate.js";

const TimelineWrapper = styled.div`
  font-size: 14px;
  line-height: 1.5;
  color: rgba(0, 0, 0, 0.65);
  list-style: none;
  margin: 0;
  padding: 0;

  &.${(props) => props.prefixCls}-wrapper-pending {
    .${(props) => props.prefixCls}-item-last {
      .${(props) => props.prefixCls}-item-tail {
        display: block;
        border-left: 2px dashed #e8e8e8;
      }
    }
  }

  &.${(props) => props.prefixCls}-wrapper-alternate {
    .${(props) => props.prefixCls}-item-left {
      .${(props) => props.prefixCls}-item-content {
        text-align: left;
        left: 50%;
        width: 50%;
      }

      .${(props) => props.prefixCls}-item-tail {
        left: 50%;
      }

      .${(props) => props.prefixCls}-item-head {
        left: 50%;
        margin-left: -4px;
      }

      .${(props) => props.prefixCls}-item-head-custom {
        margin-left: -8px;
      }
    }

    .${(props) => props.prefixCls}-item-right {
      .${(props) => props.prefixCls}-item-content {
        text-align: right;
        right: 50%;
        margin-right: 18px;
        width: 50%;
        left: -30px;
      }

      .${(props) => props.prefixCls}-item-tail {
        left: 50%;
      }

      .${(props) => props.prefixCls}-item-head {
        left: 50%;
        margin-left: -4px;
      }

      .${(props) => props.prefixCls}-item-head-custom {
        margin-left: -8px;
      }
    }
  }

  &.${(props) => props.prefixCls}-wrapper-right {
    .${(props) => props.prefixCls}-item-right {
      .${(props) => props.prefixCls}-item-content {
        right: 0;
        width: 100%;
        left: -30px;
        text-align: right;
      }

      .${(props) => props.prefixCls}-item-tail {
        left: 100%;
      }

      .${(props) => props.prefixCls}-item-head {
        left: 100%;
        margin-left: -4px;
      }

      .${(props) => props.prefixCls}-item-head-custom {
        margin-left: -8px;
      }
    }
  }
`;

const TimelineItemWrapper = styled.li`
  position: relative;
  padding: 0 0 20px;
  list-style: none;
  margin: 0;
  font-size: 14px;

  .${(props) => props.prefixCls}-item-tail {
    position: absolute;
    left: 4px;
    top: 0.85em;
    height: 100%;
    border-left: 2px solid #e8e8e8;
  }

  .${(props) => props.prefixCls}-item-head {
    position: absolute;
    width: 6px;
    height: 6px;
    background-color: #fff;
    border-radius: 100px;
    border-width: 2px;
    border-style: solid;
    border-color: ${(props) => props.color};
    color: ${(props) => props.color};
    box-sizing: content-box;

    .${(props) => props.prefixCls}-item-head-custom-icon {
      animation: ${skySpinRotate} 1s linear infinite;
    }
  }

  .${(props) => props.prefixCls}-item-head-custom {
    border: none;
    width: auto;
    height: auto;
    left: -4px;
  }

  .${(props) => props.prefixCls}-item-content {
    margin: 0 0 0 18px;
    position: relative;
    top: -6px;
  }

  &.${(props) => props.prefixCls}-item-last {
    .${(props) => props.prefixCls}-item-content {
      min-height: 48px;
    }

    .${(props) => props.prefixCls}-item-tail {
      display: none;
    }
  }

  &.${(props) => props.prefixCls}-item-pending {
    .${(props) => props.prefixCls}-item-head {
      font-size: 12px;
      width: auto;
      height: auto;
      left: -2px;
    }

    .${(props) => props.prefixCls}-item-tail {
      display: none;
    }
  }
`;

export { TimelineWrapper, TimelineItemWrapper };
