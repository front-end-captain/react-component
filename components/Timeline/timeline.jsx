import React, { Children, cloneElement } from "react";
import classNames from "classnames";
import PropTypes from "prop-types";

import TimelineItem from "./timelineItem.jsx";
import { CircleLoadingIcon } from "./../Icon/index.jsx";

import { TimelineWrapper } from "./timeline.css.js";

const Timeline = (props) => {
  const { prefixCls, style, className, pending, mode, pendingDot, children } = props;

  const timelineWrapperClassName = classNames(
    `${prefixCls}-wrapper`,
    className,
    `${prefixCls}-wrapper-${mode}`,
    {
      [`${prefixCls}-wrapper-pending`]: pending,
    },
  );

  const pendingNode = typeof pending === "boolean" ? null : pending;
  const pendingItem = pending ? (
    <TimelineItem
      pending={!!pending}
      dot={pendingDot || <CircleLoadingIcon className={`${prefixCls}-item-head-custom-icon`} />}
    >
      {pendingNode}
    </TimelineItem>
  ) : null;

  // const timelineItems = reverse
  //   ? [pendingItem, ...Children.toArray(children).reverse()]
  //   : [...Children.toArray(children), pendingItem];

  const timelineItems = [...Children.toArray(children), pendingItem];
  const realItems = timelineItems.filter((item) => item !== null);
  const itemsAmount = Children.count(realItems);

  const item = Children.map(realItems, (child, index) => {
    const last = pending ? index === itemsAmount - 2 : index === itemsAmount - 1;
    const left = (mode === "alternate" && index % 2 === 0) || mode === "left";
    const right = mode === "right" || !left;
    const itemClassName = classNames(child.props.className, {
      [`${prefixCls}-item-last`]: last,
      [`${prefixCls}-item-left`]: left,
      [`${prefixCls}-item-right`]: right,
    });

    return cloneElement(child, { className: itemClassName });
  });

  return (
    <TimelineWrapper className={timelineWrapperClassName} style={style} prefixCls={prefixCls}>
      {item}
    </TimelineWrapper>
  );
};

Timeline.defaultProps = {
  prefixCls: "sky-timeline",
  pending: false,
  // reverse: false,
  mode: "left",
};

Timeline.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  pending: PropTypes.oneOfType([PropTypes.bool, PropTypes.element, PropTypes.string]),
  style: PropTypes.objectOf(PropTypes.string),
  // reverse: PropTypes.bool,
  mode: PropTypes.oneOf(["left", "alternate", "right"]),
};

Timeline.Item = TimelineItem;

export default Timeline;
