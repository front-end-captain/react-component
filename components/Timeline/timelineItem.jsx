import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import { TimelineItemWrapper } from "./timeline.css.js";

const TimelineItem = (props) => {
  const { prefixCls, style, className, color, dot, pending, children } = props;

  const itemClassName = classNames(`${prefixCls}-item`, className, {
    [`${prefixCls}-item-pending`]: pending,
  });

  const itemDotClassName = classNames(`${prefixCls}-item-head`, {
    [`${prefixCls}-item-head-custom`]: dot,
  });

  return (
    <TimelineItemWrapper
      className={itemClassName}
      style={style}
      color={color}
      prefixCls={prefixCls}
    >
      <div className={`${prefixCls}-item-tail`} />
      <div className={itemDotClassName}>{dot}</div>
      <div className={`${prefixCls}-item-content`}>{children}</div>
    </TimelineItemWrapper>
  );
};

TimelineItem.defaultProps = {
  prefixCls: "sky-timeline",
  color: "blue",
  pending: false,
};

TimelineItem.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  color: PropTypes.string,
  dot: PropTypes.element,
  pending: PropTypes.oneOfType([PropTypes.bool, PropTypes.element, PropTypes.string]),
  style: PropTypes.objectOf(PropTypes.string),
};

export default TimelineItem;
