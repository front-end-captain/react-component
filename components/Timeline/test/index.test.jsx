import React from "react";
import { render } from "enzyme";
import toJson from "enzyme-to-json";

import Timeline from "./../index.js";

import { InfoIcon } from "./../../Icon/index.jsx";

const { Item } = Timeline;

describe("Timeline", () => {
  it("should render some Timeline components", () => {
    const wrapper = render(
      <div>
        <Timeline>
          <Item>Create a services site 2015-09-01</Item>
          <Item>Solve initial network problems 2015-09-01</Item>
          <Item>Technical testing 2015-09-01</Item>
          <Item>Network problems being solved 2015-09-01</Item>
        </Timeline>
        <Timeline pending="Recording...">
          <Item>Create a services site 2015-09-01</Item>
          <Item>Solve initial network problems 2015-09-01</Item>
          <Item>Technical testing 2015-09-01</Item>
        </Timeline>
        <Timeline mode="alternate">
          <Item>Create a services site 2015-09-01</Item>
          <Item color="green">Solve initial network problems 2015-09-01</Item>
          <Item dot={<InfoIcon style={{ color: "red", fontSize: "18px" }} />}>
            Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque
            laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi
            architecto beatae vitae dicta sunt explicabo.
          </Item>
          <Item color="red">Network problems being solved 2015-09-01</Item>
          <Item>Create a services site 2015-09-01</Item>
          <Item dot={<InfoIcon style={{ color: "red", fontSize: "18px" }} />}>
            Technical testing 2015-09-01
          </Item>
        </Timeline>
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
