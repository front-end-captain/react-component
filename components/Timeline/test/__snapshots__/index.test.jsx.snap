// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline should render some Timeline components 1`] = `
<div>
  <div
    class="sky-timeline-wrapper sky-timeline-wrapper-left timelinecss__TimelineWrapper-sc-1vumii2-0 BbkAW"
  >
    <li
      class="sky-timeline-item sky-timeline-item-left timelinecss__TimelineItemWrapper-sc-1vumii2-1 ecnAVp"
      color="blue"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head"
      />
      <div
        class="sky-timeline-item-content"
      >
        Create a services site 2015-09-01
      </div>
    </li>
    <li
      class="sky-timeline-item sky-timeline-item-left timelinecss__TimelineItemWrapper-sc-1vumii2-1 ecnAVp"
      color="blue"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head"
      />
      <div
        class="sky-timeline-item-content"
      >
        Solve initial network problems 2015-09-01
      </div>
    </li>
    <li
      class="sky-timeline-item sky-timeline-item-left timelinecss__TimelineItemWrapper-sc-1vumii2-1 ecnAVp"
      color="blue"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head"
      />
      <div
        class="sky-timeline-item-content"
      >
        Technical testing 2015-09-01
      </div>
    </li>
    <li
      class="sky-timeline-item sky-timeline-item-last sky-timeline-item-left timelinecss__TimelineItemWrapper-sc-1vumii2-1 ecnAVp"
      color="blue"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head"
      />
      <div
        class="sky-timeline-item-content"
      >
        Network problems being solved 2015-09-01
      </div>
    </li>
  </div>
  <div
    class="sky-timeline-wrapper sky-timeline-wrapper-left sky-timeline-wrapper-pending timelinecss__TimelineWrapper-sc-1vumii2-0 BbkAW"
  >
    <li
      class="sky-timeline-item sky-timeline-item-left timelinecss__TimelineItemWrapper-sc-1vumii2-1 ecnAVp"
      color="blue"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head"
      />
      <div
        class="sky-timeline-item-content"
      >
        Create a services site 2015-09-01
      </div>
    </li>
    <li
      class="sky-timeline-item sky-timeline-item-left timelinecss__TimelineItemWrapper-sc-1vumii2-1 ecnAVp"
      color="blue"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head"
      />
      <div
        class="sky-timeline-item-content"
      >
        Solve initial network problems 2015-09-01
      </div>
    </li>
    <li
      class="sky-timeline-item sky-timeline-item-last sky-timeline-item-left timelinecss__TimelineItemWrapper-sc-1vumii2-1 ecnAVp"
      color="blue"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head"
      />
      <div
        class="sky-timeline-item-content"
      >
        Technical testing 2015-09-01
      </div>
    </li>
    <li
      class="sky-timeline-item sky-timeline-item-left sky-timeline-item-pending timelinecss__TimelineItemWrapper-sc-1vumii2-1 ecnAVp"
      color="blue"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head sky-timeline-item-head-custom"
      >
        <svg
          class="sky-timeline-item-head-custom-icon"
          fill="currentColor"
          height="1em"
          version="1.1"
          viewBox="0 0 1024 1024"
          width="1em"
          xlink="http://www.w3.org/1999/xlink"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 0 0-94.3-139.9 437.71 437.71 0 0 0-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3 0.1 19.9-16 36-35.9 36z"
            p-id="1345"
          />
        </svg>
      </div>
      <div
        class="sky-timeline-item-content"
      >
        Recording...
      </div>
    </li>
  </div>
  <div
    class="sky-timeline-wrapper sky-timeline-wrapper-alternate timelinecss__TimelineWrapper-sc-1vumii2-0 BbkAW"
  >
    <li
      class="sky-timeline-item sky-timeline-item-left timelinecss__TimelineItemWrapper-sc-1vumii2-1 ecnAVp"
      color="blue"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head"
      />
      <div
        class="sky-timeline-item-content"
      >
        Create a services site 2015-09-01
      </div>
    </li>
    <li
      class="sky-timeline-item sky-timeline-item-right timelinecss__TimelineItemWrapper-sc-1vumii2-1 ceYtjC"
      color="green"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head"
      />
      <div
        class="sky-timeline-item-content"
      >
        Solve initial network problems 2015-09-01
      </div>
    </li>
    <li
      class="sky-timeline-item sky-timeline-item-left timelinecss__TimelineItemWrapper-sc-1vumii2-1 ecnAVp"
      color="blue"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head sky-timeline-item-head-custom"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color:red;font-size:18px"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M235.4 172.2c0-11.4 9.3-19.9 20.5-19.9 11.4 0 20.7 8.5 20.7 19.9s-9.3 20-20.7 20c-11.2 0-20.5-8.6-20.5-20zm1.4 35.7H275V352h-38.2V207.9z"
          />
          <path
            d="M256 76c48.1 0 93.3 18.7 127.3 52.7S436 207.9 436 256s-18.7 93.3-52.7 127.3S304.1 436 256 436c-48.1 0-93.3-18.7-127.3-52.7S76 304.1 76 256s18.7-93.3 52.7-127.3S207.9 76 256 76m0-28C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48z"
          />
        </svg>
      </div>
      <div
        class="sky-timeline-item-content"
      >
        Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo.
      </div>
    </li>
    <li
      class="sky-timeline-item sky-timeline-item-right timelinecss__TimelineItemWrapper-sc-1vumii2-1 dOfllj"
      color="red"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head"
      />
      <div
        class="sky-timeline-item-content"
      >
        Network problems being solved 2015-09-01
      </div>
    </li>
    <li
      class="sky-timeline-item sky-timeline-item-left timelinecss__TimelineItemWrapper-sc-1vumii2-1 ecnAVp"
      color="blue"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head"
      />
      <div
        class="sky-timeline-item-content"
      >
        Create a services site 2015-09-01
      </div>
    </li>
    <li
      class="sky-timeline-item sky-timeline-item-last sky-timeline-item-right timelinecss__TimelineItemWrapper-sc-1vumii2-1 ecnAVp"
      color="blue"
    >
      <div
        class="sky-timeline-item-tail"
      />
      <div
        class="sky-timeline-item-head sky-timeline-item-head-custom"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color:red;font-size:18px"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M235.4 172.2c0-11.4 9.3-19.9 20.5-19.9 11.4 0 20.7 8.5 20.7 19.9s-9.3 20-20.7 20c-11.2 0-20.5-8.6-20.5-20zm1.4 35.7H275V352h-38.2V207.9z"
          />
          <path
            d="M256 76c48.1 0 93.3 18.7 127.3 52.7S436 207.9 436 256s-18.7 93.3-52.7 127.3S304.1 436 256 436c-48.1 0-93.3-18.7-127.3-52.7S76 304.1 76 256s18.7-93.3 52.7-127.3S207.9 76 256 76m0-28C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48z"
          />
        </svg>
      </div>
      <div
        class="sky-timeline-item-content"
      >
        Technical testing 2015-09-01
      </div>
    </li>
  </div>
</div>
`;
